source/slang/slang-ir-entry-point-uniforms.cpp

```diff
@@ -549,8 +549,10 @@ struct CollectEntryPointUniformParams : PerEntryPointPass
             if (m_options.targetReq->getOptionSet().getBoolOption(
                     CompilerOptionName::GLSLForceScalarLayout))
                 layoutType = builder.getType(kIROp_ScalarBufferLayoutType);
-            else if (m_options.targetReq->getOptionSet().getBoolOption(
-                         CompilerOptionName::ForceCLayout))
+            else if (
+                m_options.targetReq->getOptionSet().getBoolOption(
+                    CompilerOptionName::ForceCLayout) ||
+                isLLVMTarget(m_options.targetReq))
```

Comment:
Again, I don't like this idea of forcing things to use C-style layout rules just because of the choice to compile to LLVM.

We should have an actual discussion about what the layout rules for a CPU target *should* be, and then implement those rules consistently (ideally for both the old and new CPU back-ends, but I accept that there could be divergence if the old back-end can't easily be modified).

I continue to believe that C-style layout is something people should be able to ask for on a type, but that they should *have* to ask for it, explicitly. (Any bridge code generator could, of course, attach attributes or whatever to types to request C-style layout for them, so that people using imported C APIs wouldn't need to worry about it)

Reply:
I think it would require quite a bit of effort to try to support arbitrary layout rules (that aren't the C-style one) on the old CPU target. It currently simply emits C++ structs without extra padding fields and no packing. Still, I'm open to basically any layout here, as long as it's possible to somehow enforce C-style layout for bindings of C APIs as you mentioned.

I'll try and fix these forced layout rules during the weekend.